ecr/ref.go

```diff
@@ -70,6 +74,60 @@ func (spec ECRSpec) Registry() string {
 	return spec.arn.AccountID
 }
 
+// ParseEcrImageNameToRef takes an ECR image name and returns its canonical `ref` format.
+// i.e. ecr.aws/arn:<partition>:ecr:<region>:<account>:repository/<name>:<tag>
+func ParseEcrImageNameToRef(input string) (string, error) {
+	ref := "ecr.aws/"
+	partition := "aws"
+	if strings.HasPrefix(input, "https://") {
+		input = strings.TrimPrefix(input, "https://")
+	}
+	// Matching on account, region and TLD
+	err := errors.New("Invalid ECR image name")
+	matches := ecrRegex.FindStringSubmatch(input)
+	if len(matches) < 3 {
+		return "", err
+	}
+	tld := matches[3]
+	region := matches[2]
+	account := matches[1]
+	// If `.cn` TLD, partition should be "aws-cn"
+	// If US gov cloud regions, partition should be "aws-us-gov"
+	// If both of them match, the image source is invalid
+	isCnEndpoint := tld == ".cn"
+	isGovCloudEndpoint := region == "us-gov-west-1" || region == "us-gov-east-1"
+	if isCnEndpoint && isGovCloudEndpoint {
+		return "", err
+	} else if isCnEndpoint {
+		partition = "aws-cn"
+	} else if isGovCloudEndpoint {
+		partition = "aws-us-gov"
+	}
+	// Separate out <name>:<tag> for checking validity
+	tokens := strings.Split(input, "/")
+	if len(tokens) < 2 {
+		return "", errors.New("No specified name and tag or digest")
+	}
+	fullImageId := tokens[len(tokens)-1]
+	matchDigest, _ := regexp.MatchString(`^[a-zA-Z0-9-_]+@sha256:[A-Fa-f0-9]{64}$`, fullImageId)
+	matchTag, _ := regexp.MatchString(`^[a-zA-Z0-9-_]+:[a-zA-Z0-9\.\-_]{1,128}$`, fullImageId)
```

Comment:
There's logic for this in the `parseARN` function already (look at how it populates the `object` variable).  I'd recommend refactoring that out and using it in both places.

Reply:
It looks like you factored the logic out of `parseARN`, but you're not using it here yet.  Can you use it here too?

Reply:
I use these regex to check for the validity of the tag/digest. The logic in `parseARN` is concerned with separating the repository and the object (tag or digest) and does not check if the tag/digest is valid.

Reply:
Why do you want to try and validate them?

Reply:
I think this is probably the wrong place to do validation.  You'll get meaningful errors elsewhere.  The digest right now is only `sha256`, but ECR and containerd *could* both gain support for other digests.  The tag length limit is an ECR limit and should be enforced by the ECR service.  Client-side validation like this tightly couples the library to the current implementation details of the things it's integrating with, instead of allowing those components to enforce their own validation.

Reply:
If an image "URI" with a non-existent/invalid tag/digest was passed, the resultant `ECRSpec` will have an empty string for the `object` which would cause any attempts to pull/push image fail. Which I guess is fine? since it'll error out down the line.

Reply:
I'll go ahead and remove these lines.

Reply:
It should end up with something in `Object` as long as something was provided after `:` or `@`, but ECR should ultimately be the arbiter of "is it valid".